File: LaRC/g3b.ssp.2022062949SSv05.21.py
"""
This example code illustrates how to access and visualize a LaRC ASDC g3bssp
vertical profile HDF5 file in Python.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python g3b.ssp.2022062949SSv05.21.py

The HDF5 file must be in your current working directory.

Tested under: Python 3.9.12 :: Miniconda
Last updated: 2022-08-25
"""

import os
import h5py

import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt

from matplotlib.ticker import FormatStrFormatter
from matplotlib.ticker import ScalarFormatter

FILE_NAME = 'g3b.ssp.2022062949SSv05.21'

with h5py.File(FILE_NAME, mode='r') as f:

    dset_var = f['/Altitude Based Data/Aerosol Ozone Profiles/Ozone_AO3']
    dset_pres = f['/Altitude Based Data/Altitude Information/Altitude']
    dset_time = f['/Event Information/Time']
    dset_date = f['/Event Information/Date']
    dset_lat = f['/Event Information/Latitude']
    dset_lon = f['/Event Information/Longitude']

    # Read the data.
    data = dset_var[:]
    altitude = dset_pres[:]
    time = dset_time[()]
    dt = dset_date[()]
    lat = dset_lat[()]
    lon = dset_lon[()]
    
    # Read the needed attributes.
    data_units = dset_var.attrs['units']
    pres_units = dset_pres.attrs['units']
    data_title = dset_var.attrs['long_name']
    time_title = dset_time.attrs['long_name']
    pres_title = dset_pres.attrs['long_name']

    fillvalue = dset_var.attrs['_FillValue']
    data[data == fillvalue] = np.nan

    # The date is stored as a six-digit number, YYYYMM.  Convert it into
    # a string.
    dstr = str(dt)
    ymd_str = dstr[0:4] + "-" + dstr[4:6] + "-" + dstr[6:8]

    # The time is stored as a six-digit number, HHMMSS.  Convert it into
    # a string.    
    tstr = str(time).zfill(6)
    hms_str = tstr[0:2] + ":" + tstr[2:4] + ":" + tstr[4:6]
    datestr =  ymd_str + "T" + hms_str + "Z"

    # Set lat/lon string.
    locstr = "Longitude = " + str(lon) + ", Latitude = " + str(lat)
    
    plt.plot(data, altitude)

    # Save some screen space by using scientific notation for the xtick labels.
    formatter = plt.ScalarFormatter(useMathText=True)
    formatter.set_scientific(True)
    formatter.set_powerlimits((-3, 4))
    plt.gca().xaxis.set_major_formatter(formatter)

    plt.xlabel('{0} ({1})'.format(data_title, data_units))
    plt.ylabel('{0} ({1})'.format(pres_title, pres_units))

    basename = os.path.basename(FILE_NAME)
    plt.title('{0}\n{1} at {2}\n{3}'.format(basename, data_title, datestr, locstr))

    # This is useful for putting high pressure at the bottom.
    # plt.gca().invert_yaxis()

    # Use log scale.
    # plt.gca().set_yscale('log')

    # %g will take a number that could be represented as %f (a simple float or
    # double) or %e (scientific notation) and return it as the shorter of the
    # two.
    plt.gca().yaxis.set_major_formatter(FormatStrFormatter('%g'))
    
    fig = plt.gcf()
    
    pngfile = "{0}.py.png".format(basename)    
    fig.savefig(pngfile)

File: LaRC/g3b_ssp_2022062949SSv05_21.m
%
%  This example code illustrates how to access and visualize
% LaRC ASDC g3bssp vertical profile HDF5 file in MATLAB. 
%
%  If you have any questions, suggestions, comments  on this example, 
% pleaseuse the HDF-EOS Forum (http://hdfeos.org/forums). 
%
%  If you would like to see an  example of any other NASA
% HDF/HDF-EOS data product that is not listed in the
% HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), 
% feel free to contact us at eoshelp@hdfgroup.org or post it at the
% HDF-EOS Forum (http://hdfeos.org/forums).
%
% You can run this script in batch mode as follows:
%
% $matlab -nosplash -nodesktop -r g3b_ssp_2022062949SSv05_21
%
% Please note that there is no .m in the above command at the end.
%
% Tested Under:
% Tested under: MATLAB R2021a
% Last Update: 2022-09-09

% Open the HDF5 File.
FILE_NAME = 'g3b.ssp.2022062949SSv05.21';
file_id = H5F.open(FILE_NAME, 'H5F_ACC_RDONLY', 'H5P_DEFAULT');

% Read the O3 dataset.
DATAFIELD_NAME = '/Altitude Based Data/Aerosol Ozone Profiles/Ozone_AO3';
data_id = H5D.open(file_id, DATAFIELD_NAME);
data = H5D.read(data_id,'H5T_NATIVE_DOUBLE', 'H5S_ALL', 'H5S_ALL', ...
                'H5P_DEFAULT');

% Read the altitude dataset.
ALTITUDE_NAME = '/Altitude Based Data/Altitude Information/Altitude';
alt_id = H5D.open(file_id, ALTITUDE_NAME);
altitude = H5D.read(alt_id,'H5T_NATIVE_DOUBLE', 'H5S_ALL', 'H5S_ALL',...
                  'H5P_DEFAULT');

% Read the time dataset. 
TIME_NAME = '/Event Information/Time';
time_id = H5D.open(file_id, TIME_NAME);
time = H5D.read(time_id,'H5T_NATIVE_DOUBLE', 'H5S_ALL', 'H5S_ALL',...
                'H5P_DEFAULT');

% Read the date dataset. 
DATE_NAME = '/Event Information/Date';
date_id = H5D.open(file_id, DATE_NAME);
date = H5D.read(date_id,'H5T_NATIVE_DOUBLE', 'H5S_ALL', 'H5S_ALL',...
              'H5P_DEFAULT');

% Read the lat dataset. 
LAT_NAME = '/Event Information/Latitude';
lat_id = H5D.open(file_id, LAT_NAME);
lat = H5D.read(lat_id,'H5T_NATIVE_DOUBLE', 'H5S_ALL', 'H5S_ALL',...
              'H5P_DEFAULT');

% Read the lon dataset. 
LON_NAME = '/Event Information/Longitude';
lon_id = H5D.open(file_id, LON_NAME);
lon = H5D.read(lon_id,'H5T_NATIVE_DOUBLE', 'H5S_ALL', 'H5S_ALL',...
              'H5P_DEFAULT');

% Read the units of data.
ATTRIBUTE = 'units';
attr_id = H5A.open_name (data_id, ATTRIBUTE);
units = H5A.read(attr_id, 'H5ML_DEFAULT');


% Read the fill value.
ATTRIBUTE = '_FillValue';
attr_id = H5A.open_name (data_id, ATTRIBUTE);
fillvalue = H5A.read (attr_id, 'H5T_NATIVE_DOUBLE');

% Read the long_name attribute.
ATTRIBUTE = 'long_name';
attr_id = H5A.open_name (data_id, ATTRIBUTE);
long_name = H5A.read(attr_id, 'H5ML_DEFAULT');

% Read the units of altitude axis.
ATTRIBUTE = 'units';
attr_id = H5A.open_name (alt_id, ATTRIBUTE);
units_alt = H5A.read(attr_id, 'H5ML_DEFAULT');

ATTRIBUTE = 'long_name';
attr_id = H5A.open_name (alt_id, ATTRIBUTE);
long_name_alt = H5A.read(attr_id, 'H5ML_DEFAULT');


% Close and release resources.
H5A.close (attr_id)
H5D.close (data_id);
H5D.close (alt_id);
H5D.close (time_id);
H5F.close (file_id);

% Replace the fill value with NaN.
data(data == fillvalue) = NaN;

f = figure('Name', FILE_NAME, 'visible', 'off');

plot(data, altitude);

ylabel([long_name_alt ' (' units_alt ')']);
yticks = get(gca,'ytick');
set(gca,'YTickLabel',yticks);

date_s = num2str(date);
year = date_s(1:4);
month = date_s(5:6);
day = date_s(7:8);
d = [year, '-', month, '-', day];

time_s = num2str(time, '%06d');
hh = time_s(1:2);
mm = time_s(3:4);
ss = time_s(5:6);
t = [hh, ':', mm, ':', ss, 'Z'];

dt = [d 'T' t];
xtitle = [long_name ' at ' dt];
xlabel([xtitle ' (' units ')']);

ax = gca;

title({FILE_NAME; ['Longitude=', num2str(lon), ...
                   ' Latitude=', num2str(lat)]}, ... 
      'Interpreter', 'None', 'FontSize',16,'FontWeight','bold');

saveas(f, [FILE_NAME, '.m.png']);
exit;
